### PR TITLE
Workflow editor UI touch up

### LINKED ISSUE #1719: Improve "Collection Type" selection in workflow UI
When adding/modifying an "Input dataset collection", "Collection Type" on the right panel is a text field on Firefox.

screenshot from 2016-02-12 16 28 24

Can we help the user with a multiple choice select (`list`, `paired`, `list:paired`, `other`)? Or at least add some help? The only problem may be the fact that potentially a user may want to use a `list:list` or `list:list:paired` or something even crazier.

Ping @jmchilton for suggestions.

### PR DESCRIPTION
**Pull request by guerler**
Follow-up for #3181.

1. Fixes #1719 

2. Fixes the label handling and duplicate message display

3. Moves the workflow storage endpoint from controller to api

4. Introduces better model handling for forms and improves the tool model building process

5. Adds browser-independent datalist handling using jquery's autocomplete feature

newdlist

### CODE REVIEW

**Comment by jmchilton** [comment]
- I like the datalist drop down definitely. Very clear step forward and much needed. (You should include in the PR description "Fixes #1719" :smile:).
- I like the previous editable title since that is how we want it to show up in the run form step title box and in the box title on the workflow editor. I think we could display the tool name and version on mouseover or something. But I do understand the value of the label being permanently labelled as non-unique though - that red error state indication is important and currently unimplemented. So I think this is fine with this I guess. It is a bit more basic but the utility is clear.

**Comment by guerler** [comment]
Cool. Awesome. Ill add the label handling for tools tmw then.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 1]
```diff
@@ -296,6 +296,37 @@ def import_new_workflow_deprecated(self, trans, payload, **kwd):
...
+                trans,
+                stored,
+                payload,
+            )
+        except workflows.MissingToolsException as e:
+            return dict(
+                name=e.workflow.name,
+                message="This workflow includes missing or invalid tools. It cannot be saved until the following steps are removed or the missing tools are enabled.",
+                errors=e.errors,
+            )
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: I don't think returning a `dict` here follows Galaxy tradition of raising an Exception if an API call fails, I'd rather remove the except block.

Comment (jmchilton) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: Additionally we already have an endpoint that does something very similar to this and it does it with a more RESTful URL if I recall correctly. Can we avoid duplicating this?

File: lib/galaxy/webapps/galaxy/buildapp.py [hunk 2]
```diff
@@ -284,6 +284,7 @@ def populate_api_routes( webapp, app ):
     webapp.mapper.connect( '/api/genomes/{id}/sequences', controller='genomes', action='sequences' )
     webapp.mapper.resource( 'visualization', 'visualizations', path_prefix='/api' )
     webapp.mapper.connect( '/api/workflows/build_module', action='build_module', controller="workflows" )
+    webapp.mapper.connect( '/api/workflows/{id}/save', action='save', controller="workflows" )
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/buildapp.py]: This can be removed now.

**Pull request merged**